modules/actions/snapshots/sync-snapshots.ts

```diff
@@ -0,0 +1,123 @@
+    getMetadata: () => Promise<{ block: { number: number } }>;
+    getAllSnapshots: (args: PoolSnapshot_Filter) => Promise<Prisma.PrismaPoolSnapshotUncheckedCreateInput[]>;
+};
+
+export async function syncSnapshots(
+    subgraphClient: SnapshotsSubgraphClient,
+    category: PrismaLastBlockSyncedCategory,
+    chain: Chain,
+    options: {
+        poolIds?: string[]; // Optional: Specific pool IDs to sync
+        startFromLastSyncedBlock?: boolean; // Optional: Whether to sync based on the latest block
+    } = {},
+): Promise<string[]> {
+    const { poolIds = [], startFromLastSyncedBlock = true } = options;
+    const currentBlock = await subgraphClient.getMetadata().then((meta) => meta.block.number);
+
+    let lastBlock = 0;
+    if (startFromLastSyncedBlock) {
+        lastBlock = await prisma.prismaLastBlockSynced
+            .findFirst({
+                where: { chain, category },
+                select: { blockNumber: true },
+            })
+            .then((lastBlock) => lastBlock?.blockNumber || 0);
+    }
+
+    // Fetch pool IDs from DB
+    const poolIdsInDb = await prisma.prismaPool
+        .findMany({ where: { chain, ...(poolIds.length > 0 ? { id: { in: poolIds } } : {}) } })
+        .then((pools) => pools.map((pool) => pool.id));
+
+    if (poolIdsInDb.length === 0) {
+        return [];
+    }
+
+    // Fetch snapshots from subgraph client
+    const snapshots = await subgraphClient
+        .getAllSnapshots({
+            pool_: { isInitialized: true },
+            ...(startFromLastSyncedBlock && lastBlock > 0 ? { _change_block: { number_gte: lastBlock } } : {}),
+            ...(poolIds.length > 0 ? { pool_in: poolIds } : {}),
+        })
+        .then((snapshots) => snapshots.filter((snapshot) => poolIdsInDb.includes(snapshot.poolId)));
+
+    if (!snapshots || snapshots.length === 0) {
+        return [];
+    }
+
+    let mergedSnapshots: Prisma.PrismaPoolSnapshotUncheckedCreateInput[] = snapshots;
+
+    if (startFromLastSyncedBlock) {
+        // Get pool IDs and fetch the 2 latest snapshots for each from the DB
+        const poolIds = [...new Set(snapshots.map((snapshot) => snapshot.poolId))];
+        const dbSnapshots: Prisma.PrismaPoolSnapshotUncheckedCreateInput[] = await prisma.$queryRaw<
+            (Prisma.PrismaPoolSnapshotUncheckedCreateInput & { row_num: number })[]
+        >`
+            WITH latest_snapshots AS (
+                SELECT *,
+                    ROW_NUMBER() OVER (PARTITION BY "poolId" ORDER BY "timestamp" DESC) AS row_num
+                FROM "PrismaPoolSnapshot"
+                WHERE "chain" = ${chain}::"Chain" AND "poolId" IN (${Prisma.join(poolIds)})
+            )
+            SELECT *
+            FROM latest_snapshots
+            WHERE row_num <= 2;
+        `.then((snapshots) => snapshots.map(({ row_num, ...snapshot }) => snapshot));
+
+        // Merge the new snapshots with the existing ones
+        mergedSnapshots = Object.values(
```

Comment:
personally, I find this very hard to read. I'd prefer a more verbose merging here without reduce.